bot/exts/moderation/clean.py

```diff
@@ -0,0 +1,598 @@
+                channels = [second_limit.channel]
+            else:
+                channels = [ctx.channel]
+
+        if isinstance(first_limit, Message):
+            first_limit = first_limit.created_at
+        if isinstance(second_limit, Message):
+            second_limit = second_limit.created_at
+        if first_limit and second_limit:
+            first_limit, second_limit = sorted([first_limit, second_limit])
+
+        # Needs to be called after standardizing the input.
+        predicate = self._build_predicate(bots_only, users, regex, first_limit, second_limit)
+
+        # Delete the invocation first
+        await self._delete_invocation(ctx)
+
+        if channels == "*" and use_cache:
+            message_mappings, message_ids = self._get_messages_from_cache(traverse=traverse, to_delete=predicate)
+        else:
+            deletion_channels = channels
+            if channels == "*":
+                deletion_channels = [channel for channel in ctx.guild.channels if isinstance(channel, TextChannel)]
+            message_mappings, message_ids = await self._get_messages_from_channels(
+                traverse=traverse,
+                channels=deletion_channels,
+                to_delete=predicate,
+                before=second_limit,
+                after=first_limit  # Remember first is the earlier datetime.
+            )
+
+        if not self.cleaning:
+            # Means that the cleaning was canceled
+            return
+
+        # Now let's delete the actual messages with purge.
+        self.mod_log.ignore(Event.message_delete, *message_ids)
+        deleted_messages = await self._delete_found(message_mappings)
+        self.cleaning = False
+
+        logged = await self._modlog_cleaned_messages(deleted_messages, channels, ctx)
+
+        if logged and is_mod_channel(ctx.channel):
+            with suppress(NotFound):  # Can happen if the invoker deleted their own messages.
+                await ctx.message.add_reaction(Emojis.check_mark)
+
+    # region: Commands
+
+    @group(invoke_without_command=True, name="clean", aliases=["clear", "purge"])
+    async def clean_group(
+            self,
+            ctx: Context,
+            users: Greedy[User] = None,
+            traverse: Optional[int] = None,
+            first_limit: Optional[CleanLimit] = None,
+            second_limit: Optional[CleanLimit] = None,
+            use_cache: Optional[bool] = None,
+            bots_only: Optional[bool] = False,
+            regex: Optional[Regex] = None,
+            *,
+            channels: CleanChannels = None  # "Optional" with discord.py silently ignores incorrect input.
+    ) -> None:
+        """
+        Commands for cleaning messages in channels.
+
+        If arguments are provided, will act as a master command from which all subcommands can be derived.
+        • `users`: A series of user mentions, ID's, or names.
+        • `traverse`: The number of messages to look at in each channel.
```

Comment:
I don't believe this is currently accurate if `_get_messages_from_cache` is used (e.g. in the case of cleaning from all channels), as in that case it will just look through the past `traverse` messages in the cache rather than per channel.